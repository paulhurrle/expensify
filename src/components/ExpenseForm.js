import React from 'react';
import moment from 'moment'; // go to moment.js
import { SingleDatePicker } from 'react-dates'; // go to github.com/airbnb/react-dates

export default class ExpenseForm extends React.Component {
  constructor(props) {
    super(props);

    this.state = {
      description: props.expense ? props.expense.description : '',
      note: props.expense ? props.expense.note : '',
      amount: props.expense ? (props.expense.amount / 100).toString() : '',
      createdAt: props.expense ? moment(props.expense.createdAt) : moment(),
      calendarFocused: false,
      error: '',
    };
  }
  
  onDescriptionChange = (e) => {
    const description = e.target.value;
    this.setState(() => ({ description }));
  };
  onNoteChange = (e) => {
    const note = e.target.value;
    this.setState(() => ({ note }));
  };
  onAmountChange = (e) => {
    const amount = e.target.value;
    if (!amount || amount.match(/^\d{1,}(\.\d{0,2})?$/)) { // allows user to clear the amount field and set state just as well as type in a value that matches the regex
      this.setState(() => ({ amount }));
    }
  };
  onDateChange = (createdAt) => {
    if (createdAt) { // if a value exists, set state, but do not allow an empty value (e.g. delete)
      this.setState(() => ({ createdAt }));
    }
  };
  onFocusChange = ({ focused }) => {
    this.setState(() => ({ calendarFocused: focused }));
  };
  onSubmit = (e) => {
    e.preventDefault(); // prevent a full page refresh
    if (!this.state.description || !this.state.amount) {
      this.setState(() => ({
        error: 'Please include a description and an amount.'
      }));
    } else {
      this.setState(() => ({ error: '' }));
      this.props.onSubmit({
        description: this.state.description,
        amount: parseFloat(this.state.amount, 10) * 100, // translates amount string into a decimal number
        createdAt: this.state.createdAt.valueOf(), // valueOf comes from moment API
        note: this.state.note,
      });
    }
  };
  render() {
    return (
      <div>
        {this.state.error.length > 0 && <p>{this.state.error}</p>}
        <form onSubmit={this.onSubmit}>
          <input
            type="text"
            placeholder="Description"
            autoFocus
            value={this.state.description}
            onChange={this.onDescriptionChange}
          />
          <input
            type="text"
            placeholder="Amount"
            value={this.state.amount}
            onChange={this.onAmountChange}
          />
          <SingleDatePicker
            date={this.state.createdAt}
            onDateChange={this.onDateChange}
            focused={this.state.calendarFocused}
            onFocusChange={this.onFocusChange}
            numberOfMonths={1}
            isOutsideRange={() => false}
          />
          <textarea
            placeholder="Add a note for your expense (optional)"
            value={this.state.note}
            onChange={this.onNoteChange}
          >
          </textarea>
          <button>Add Expense</button>
        </form>
      </div>
    );
  };
};