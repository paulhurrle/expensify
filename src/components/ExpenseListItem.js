import React from 'react';
import { NavLink } from 'react-router-dom';
import moment from 'moment';
import numeral from 'numeral'; // converts numbers to a desired format

const ExpenseListItem = ({ description, amount, createdAt, id }) => (
  <div>
    <NavLink to={`/edit/${id}/`}>
      <h3>{description}</h3>
    </NavLink>
    <ul>
      <li key={`${description}-${amount}`}>Amount: {numeral(amount / 100).format('$0,0.00')}</li>
      <li key={`${description}-${createdAt}`}>Created: {moment(createdAt).format('MMMM Do, YYYY')}</li>
    </ul>
  </div>
);

export default ExpenseListItem;