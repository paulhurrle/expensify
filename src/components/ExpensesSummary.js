import React, { Component } from 'react';
import pluralize from 'pluralize';
import { connect } from 'react-redux';
import numeral from 'numeral';
import selectExpenses from '../selectors/expenses';
import selectExpensesTotal from '../selectors/expenses-total';

export class ExpensesSummary extends Component {
  render() {
    const { expenseCount, expensesTotal } = this.props;

    return (
      <div>
        <h3>Viewing {expenseCount} {pluralize('expense', expenseCount)} totalling {numeral(expensesTotal / 100).format('$0,0.00')}</h3>
      </div>
    );
  };
};

const mapStateToProps = (state) => {
  const visibleExpenses = selectExpenses(state.expenses, state.filters);

  return {
    expenseCount: visibleExpenses.length,
    expensesTotal: selectExpensesTotal(visibleExpenses),
  };
};

export default connect(mapStateToProps)(ExpensesSummary);