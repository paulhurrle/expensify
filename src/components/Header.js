import React from 'react';
import { NavLink } from 'react-router-dom';  // Link and NavLink used for client-side routing

const Header = () => (
  <header>
    <h1>Expensify</h1>
    <NavLink to="/" activeClassName="is-active" exact={true}>Dashboard</NavLink>
    <NavLink to="/create" activeClassName="is-active">Add an Expense</NavLink>
  </header>
);

export default Header;
