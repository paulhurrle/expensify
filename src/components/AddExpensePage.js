import React from 'react';
import ExpenseForm from './ExpenseForm';
import { connect } from 'react-redux';
import { addExpense } from '../actions/expenses';

export class AddExpensePage extends React.Component {
  onSubmit = (expense) => { // by passing in an onSubmit prop here instead of within the ExpenseForm component directly, it makes ExpenseForm more reusable!
    this.props.addExpense(expense);
    this.props.history.push('/'); // push is accessed from the default props available via the Route component, setting a route of '/' which in this case is 'Dashboard'
  };
  render() {
    return (
      <div>
        <h1>Add Expense</h1>
        <ExpenseForm onSubmit={this.onSubmit} />
      </div>
    );
  };
};

const mapDispatchToProps = (dispatch) => ({
  addExpense: (expense) => dispatch(addExpense(expense)),
});

export default connect(undefined, mapDispatchToProps)(AddExpensePage);