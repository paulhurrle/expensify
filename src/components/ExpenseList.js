import React from 'react';
import { connect } from 'react-redux';
import ExpenseListItem from './ExpenseListItem';
import selectExpenses from '../selectors/expenses';

export const ExpenseList = (props) => (
  <div>
    {
      props.expenses.length === 0 ? (
        <p>No expenses</p>
      ) : (
        props.expenses.map((expense) => (
          <ExpenseListItem key={expense.id} {...expense} />
        ))
      )
    }
  </div>
);

const mapStateToProps = (state) => {
  return {
    expenses: selectExpenses(state.expenses, state.filters),
  };
};

// define the higher order component here using the connect function
// connect takes a function as an argument that returns an object containing the component needs to access in the store
// export default the HOC
export default connect(mapStateToProps)(ExpenseList);