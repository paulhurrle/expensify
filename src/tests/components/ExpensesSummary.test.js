import React from 'react';
import { shallow } from 'enzyme';
import { ExpensesSummary } from '../../components/ExpensesSummary';

it('should summarize 0 expenses', () => {
  const props = {
    expenseCount: 0,
    expensesTotal: 0,
  };
  expect(shallow(<ExpensesSummary {...props} />)).toMatchSnapshot();
});

it('should summarize one expense', () => {
  const props = {
    expenseCount: 1,
    expensesTotal: 98765,
  };
  expect(shallow(<ExpensesSummary {...props} />)).toMatchSnapshot();
});

it('should summarize multiple expenses', () => {
  const props = {
    expenseCount: 8,
    expensesTotal: 123456789,
  };
  expect(shallow(<ExpensesSummary {...props} />)).toMatchSnapshot();
});