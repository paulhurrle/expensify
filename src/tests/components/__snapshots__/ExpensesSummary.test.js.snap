// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should summarize 0 expenses 1`] = `
<div>
  <h3>
    Viewing 
    0
     
    expenses
     totalling 
    $0.00
  </h3>
</div>
`;

exports[`should summarize multiple expenses 1`] = `
<div>
  <h3>
    Viewing 
    8
     
    expenses
     totalling 
    $1,234,567.89
  </h3>
</div>
`;

exports[`should summarize one expense 1`] = `
<div>
  <h3>
    Viewing 
    1
     
    expense
     totalling 
    $987.65
  </h3>
</div>
`;
